modules/rostests/apitests/ntdll/RtlCriticalSection.c

```diff
@@ -0,0 +1,374 @@
+
+    printf("ThreadProc2 waiting\n");
+    WaitForSingleObject(hEventThread2Cont, INFINITE);
+    printf("ThreadProc2 returned from wait\n");
+
+    RtlLeaveCriticalSection(&CritSect);
+
+    return 0;
+}
+
+static
+void
+Test_Acquire(void)
+{
+    DWORD dwThreadId1, dwThreadId2;
+
+    RtlInitializeCriticalSection(&CritSect);
+
+    // Acquire once
+    RtlEnterCriticalSection(&CritSect);
+    ok_long(CritSect.LockCount, -2);
+    ok_long(CritSect.RecursionCount, 1);
+    ok_ptr(CritSect.OwningThread, UlongToHandle(GetCurrentThreadId()));
+    ok_ptr(CritSect.LockSemaphore, NULL);
+    //ok_size_t(CritSect.SpinCount, 0x20007d0);
+
+    // Acquire recursively
+    RtlEnterCriticalSection(&CritSect);
+    ok_long(CritSect.LockCount, -2);
+    ok_long(CritSect.RecursionCount, 2);
+    ok_ptr(CritSect.OwningThread, UlongToHandle(GetCurrentThreadId()));
+    ok_ptr(CritSect.LockSemaphore, NULL);
+    //ok_size_t(CritSect.SpinCount, 0x20007d0);
+
+    hEventThread1Ready = CreateEvent(NULL, TRUE, FALSE, NULL);
+    hEventThread1Cont = CreateEvent(NULL, TRUE, FALSE, NULL);
+    hEventThread2Ready = CreateEvent(NULL, TRUE, FALSE, NULL);
+    hEventThread2Cont = CreateEvent(NULL, TRUE, FALSE, NULL);
+
+    // Create thread 1 and wait to it time to try to acquire the critical section
+    CreateThread(NULL, 0, ThreadProc1, NULL, 0, &dwThreadId1);
+    Sleep(10);
+
+    ok_long(CritSect.LockCount, -6);
+    ok_long(CritSect.RecursionCount, 2);
+    ok_ptr(CritSect.OwningThread, UlongToHandle(GetCurrentThreadId()));
+    //ok_ptr(CritSect.LockSemaphore, LongToPtr(-1));
+    //ok_size_t(CritSect.SpinCount, 0x20007d0);
+
+    // Create thread 2 and wait to it time to try to acquire the critical section
+    CreateThread(NULL, 0, ThreadProc2, NULL, 0, &dwThreadId2);
+    Sleep(10);
+
+    ok_long(CritSect.LockCount, -10);
+    ok_long(CritSect.RecursionCount, 2);
+    ok_ptr(CritSect.OwningThread, UlongToHandle(GetCurrentThreadId()));
+    //ok_ptr(CritSect.LockSemaphore, LongToPtr(-1));
+    //ok_size_t(CritSect.SpinCount, 0x20007d0);
+
+    RtlLeaveCriticalSection(&CritSect);
+    RtlLeaveCriticalSection(&CritSect);
+
+    // Wait until thread 1 has acquired the critical section
+    WaitForSingleObject(hEventThread1Ready, INFINITE);
+
+    ok_long(CritSect.LockCount, -6);
+    ok_long(CritSect.RecursionCount, 1);
+    ok_ptr(CritSect.OwningThread, UlongToHandle(dwThreadId1));
+    //ok_ptr(CritSect.LockSemaphore, LongToPtr(-1));
+    //ok_size_t(CritSect.SpinCount, 0x20007cf);
+
+    // Release thread 1, wait for thread 2 to acquire the critical section
+    SetEvent(hEventThread1Cont);
+    WaitForSingleObject(hEventThread2Ready, INFINITE);
+
+    ok_long(CritSect.LockCount, -2);
+    ok_long(CritSect.RecursionCount, 1);
+    ok_ptr(CritSect.OwningThread, UlongToHandle(dwThreadId2));
+    //ok_ptr(CritSect.LockSemaphore, LongToPtr(-1));
+    //ok_size_t(CritSect.SpinCount, 0x20007cf);
+
+    // Release thread 2
+    SetEvent(hEventThread2Cont);
+}
```

Comment:
Any point in waiting on the two thread handles for a "cleaner" exit?

Reply:
Not really, but I added it anyway.